libgit/repo.go

```diff
@@ -748,3 +751,80 @@ func RenameRepo(
 	}
 	return UpdateRepoMD(ctx, config, tlfHandle, newRepoFS)
 }
+
+// GCOptions describe options foe garbage collection.
+type GCOptions struct {
+	// The most loose refs we will tolerate; if there are more loose
+	// refs, we should pack them.
+	MaxLooseRefs int
+	// TODO: object min age for pruning, and object re-packing thresholds.
+}
+
+func needsGC(storer storage.Storer, options GCOptions) (
+	doPackRefs bool, numLooseRefs int, err error) {
+	numLooseRefs, err = storer.CountLooseRefs()
+	if err != nil {
+		return false, 0, err
+	}
+
+	doPackRefs = numLooseRefs > options.MaxLooseRefs
+	// TODO: check object pruning and object re-packing thresholds.
+	return doPackRefs, numLooseRefs, nil
+}
+
+// GCRepo runs garbage collection on the specified repo, if it exceeds
+// any of the thresholds provided in `options`.
+func GCRepo(
+	ctx context.Context, config libkbfs.Config, tlfHandle *libkbfs.TlfHandle,
+	repoName string, uniqID string, options GCOptions) error {
+	log := config.MakeLogger("")
+	log.CDebugf(ctx, "Checking whether GC is needed for %s/%s",
+		tlfHandle.GetCanonicalName(), repoName)
+
+	fs, _, err := getOrCreateRepoAndID(
+		ctx, config, tlfHandle, repoName, uniqID, getOnly)
+	if err != nil {
+		return err
+	}
+
+	storer, err := NewGitConfigWithoutRemotesStorer(fs)
+	if err != nil {
+		return err
+	}
+
+	doPackRefs, _, err := needsGC(storer, options)
+	if err != nil {
+		return err
+	}
+	if !doPackRefs {
+		log.CDebugf(ctx, "Skipping GC")
+		return nil
+	}
+
+	log.CDebugf(ctx, "Locking for GC")
+	f, err := fs.Create(repoGCLockFileName)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	err = f.Lock()
+	if err != nil {
+		return err
+	}
```

Comment:
Don't we need a `f.Unlock()` here? Also it seems if unlock fails because lock expired we need to throw the writes causes by `PackRefs` away right?

Reply:
The `defer f.Close()` right above this does the unlock.  But we need to propagate the error back to the caller, I'll fix that.

If the unlock fails, then the journal flush fails, and so once I fix the above issue, this call will return an error.  Then it is up to the caller to delete the journal without flushing it (usually by just returning early and not calling `waitForJournal()`.

Reply:
Sounds good. Thanks!